### PR TITLE
GCMonitor improvements

### LINKED ISSUE #6081: Warning about excessive time in GC measures it wrong
## steps

Build a large project a few times in sbt, especially with a few scala.js modules doing fastOpt, or other ways to churn GC

## problem

SBT prints warnings that defy logic like these:

```text
[warn] 246.322 seconds of the last 10 seconds were spent in garbage collection. You may want to increase the project heap size using `-Xmx` or try a different gc algorithm, e.g. `-XX:+UseG1GC`, for better performance.
```

## expectation

There aren't 246.322 seconds in 10 ;)

## notes

Would also be nice if the suggestions were smarter. I am already using a high -Xmx and G1GC

### PR DESCRIPTION
**Pull request by nafg**
Fixes #6081

### CODE REVIEW

**Comment by nafg** [comment]
Build failed:

```text
[error] (mainProj / Test / headerCheck) There are files without headers!
[error]   /home/runner/work/sbt/sbt/main/src/test/scala/sbt/internal/StressGCMonitor.scala
[error]   /home/runner/work/sbt/sbt/main/src/test/scala/sbt/internal/GCMonitorTest.scala
```

**Comment by eed3si9n** [comment]
Thanks for the contribution.

There's a task called `headerCreate` that would create the copyright header.

**Comment by nafg** [comment]
Not sure if StressGCMonitor is a worthy inclusion. I don't know much about benchmarking and profiling. It did the trick for me though (the unit tests didn't surface the issue and I didn't update them after I dug deeper via StressGCMonitor).

**Comment by eed3si9n** [comment]
> Not sure if StressGCMonitor is a worthy inclusion.

If the code helped reproduce the problem, I'd say you should keep it there until someday someone reproduces it via unit test etc.

**Comment by eed3si9n** [comment]
Also could you copy-paste a sample output in the description of this PR?

**Comment by nafg** [comment]
What do you mean by a sample output?

The program basically instantiates a GCMonitor (not the testing version)
and tries to generate garbage for a certain number of seconds. Since I know
hotspot can eliminate behaviors that have no effect but I don't know any
specifics, the way I generated garbage is by filling a huge list with
random numbers, and then calling another method to print the sum. On each
iteration it prints the time left.

**Comment by eed3si9n** [comment]
I was just wondering if you have example of how the output looks like when you run it in sbt if you were able to build a snapshot sbt.

**Comment by nafg** [comment]
Oh the updated warning text?

I actually never did the publishLocal thing, but you can get the output by
running StressGCMonitor. I can try to do it next week if need be

**Comment by nafg** [comment]
```text
sbt:sbtRoot> mainProj/Test/runMain sbt.internal.StressGCMonitor
[warn] multiple main classes detected: run 'show discoveredMainClasses' to see the list
[info] running (fork) sbt.internal.StressGCMonitor 
[info] 9 seconds left...
[info] [warn] In the last 6 seconds, 5.257 (93.5%) were spent in garbage collection. You may want to increase the project heap size using `-Xmx` or try a different gc algorithm, e.g. `-XX:+UseG1GC`, for better performance. Heap: 0.75GB free of 1.29GB, max 5.19GB
[info] [warn] In the last 12 seconds, 10.267 (90.0%) were spent in garbage collection. You may want to increase the project heap size using `-Xmx` or try a different gc algorithm, e.g. `-XX:+UseG1GC`, for better performance. Heap: 1.05GB free of 1.84GB, max 5.19GB
[info] [warn] In the last 23 seconds, 18.65 (82.3%) were spent in garbage collection. You may want to increase the project heap size using `-Xmx` or try a different gc algorithm, e.g. `-XX:+UseG1GC`, for better performance. Heap: 1.54GB free of 3.31GB, max 5.19GB
[info] [warn] In the last 18 seconds, 14.457 (82.2%) were spent in garbage collection. You may want to increase the project heap size using `-Xmx` or try a different gc algorithm, e.g. `-XX:+UseG1GC`, for better performance. Heap: 1.24GB free of 3.48GB, max 5.19GB
[success] Total time: 65 s (01:05), completed Nov 14, 2020 7:35:54 PM
```

@eed3si9n does that help?

**Comment by eed3si9n** [comment]
@nafg Yup. Thanks.

**Comment by nafg** [comment]
I'm thinking we could tweak the wording a bit:

> were spent in garbage collection.

Maybe shorten to GC?

> increase the project heap size

Maybe "the JVM heap" is shorter and less confusing?

> or try a different gc algorithm

Maybe capitalize GC. Or maybe it would be better to say "try a different garbage _collector_" (or just "a different collector").

> Heap: 1.24GB free of 3.48GB, max 5.19GB

Maybe this message should go before the suggestion part ("You may want to..."), maybe in square brackets. I think it would be less confusing that way. (If having it altogether is agreed upon.)

**Comment by nafg** [comment]
It's nice that github has a button to merge new commits from the branch into the PR, is that new? I assume it won't show up once the PR is squashed anyway

**Comment by eed3si9n** [comment]
> I'm thinking we could tweak the wording a bit:

Go for it. In addition to what you listed "You may want to increase the project heap size" could be "Consider increasing the heap to..."

File: main/src/main/scala/sbt/internal/GCMonitor.scala [hunk 1]
```diff
@@ -7,48 +7,83 @@
...
+  final val GB = 1024 * 1024 * 1024.0
+  val runtime = Runtime.getRuntime
+  def gbString(n: Long) = f"${n / GB}%.2fGB"
+
+  override protected def emitWarning(total: Long, over: Option[Long]): Unit = {
+    val totalSeconds = total / 1000.0
+    val amountMsg = over.fold(totalSeconds + " seconds") { d =>
+      "In the last " + (d / 1000.0).ceil.toInt + f" seconds, $totalSeconds (${total.toDouble / d * 100}%.1f%%)"
     }
+    val msg = s"$amountMsg were spent in garbage collection. " +
```
Comment (nafg) [inline main/src/main/scala/sbt/internal/GCMonitor.scala]:
```suggestion
    val msg = s"$amountMsg were spent in GC. " +
```

File: main/src/main/scala/sbt/internal/GCMonitor.scala [hunk 2]
```diff
@@ -7,48 +7,83 @@
...
+      "You may want to increase the project heap size using `-Xmx` or try " +
```
Comment (nafg) [inline main/src/main/scala/sbt/internal/GCMonitor.scala]:
```suggestion
      "Consider increasing the JVM heap using `-Xmx` or try " +
```

File: main/src/main/scala/sbt/internal/GCMonitor.scala [hunk 3]
```diff
@@ -7,48 +7,83 @@
...
+      "a different gc algorithm, e.g. `-XX:+UseG1GC`, for better performance. " +
```
Comment (nafg) [inline main/src/main/scala/sbt/internal/GCMonitor.scala]:
```suggestion
      "a different collector, e.g. `-XX:+UseG1GC`, for better performance. " +
```

File: main/src/main/scala/sbt/internal/GCMonitor.scala [hunk 4]
```diff
@@ -7,48 +7,83 @@
...
+    val amountMsg = over.fold(totalSeconds + " seconds") { d =>
+      "In the last " + (d / 1000.0).ceil.toInt + f" seconds, $totalSeconds (${total.toDouble / d * 100}%.1f%%)"
     }
+    val msg = s"$amountMsg were spent in GC. " +
+      "Consider increasing the JVM heap using `-Xmx` or try " +
+      "a different collector, e.g. `-XX:+UseG1GC`, for better performance. " +
+      s"Heap: ${gbString(runtime.freeMemory())} free " +
+      s"of ${gbString(runtime.totalMemory())}, " +
+      s"max ${gbString(runtime.maxMemory())}"
+    logger.warn(msg)
```
Comment (nafg) [inline main/src/main/scala/sbt/internal/GCMonitor.scala]:
```suggestion
    val msg = s"$amountMsg were spent in GC. " +
      s"[Heap: ${gbString(runtime.freeMemory())} free " +
      s"of ${gbString(runtime.totalMemory())}, " +
      s"max ${gbString(runtime.maxMemory())}] " +
      "Consider increasing the JVM heap using `-Xmx` or try " +
      "a different collector, e.g. `-XX:+UseG1GC`, for better performance." +
    logger.warn(msg)
```

**Comment by nafg** [comment]
Now output looks like this:

```text
[info] running (fork) sbt.internal.StressGCMonitor 
[info] 9 seconds left...
[info] [warn] In the last 6 seconds, 5.184 (93.3%) were spent in GC. [Heap: 0.75GB free of 1.29GB, max 5.19GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[info] [warn] In the last 13 seconds, 10.942 (89.7%) were spent in GC. [Heap: 1.05GB free of 1.85GB, max 5.19GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[info] [warn] In the last 25 seconds, 19.504 (80.5%) were spent in GC. [Heap: 1.56GB free of 3.38GB, max 5.19GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[info] [warn] In the last 18 seconds, 15.482 (88.9%) were spent in GC. [Heap: 1.19GB free of 3.42GB, max 5.19GB] Consider increasing the JVM heap using `-Xmx` or try a different collector, e.g. `-XX:+UseG1GC`, for better performance.
[success] Total time: 87 s (01:27), completed Nov 15, 2020 12:06:39 AM
```

**Review by eed3si9n — APPROVED** [review]

**Comment by SethTisue** [comment]
@nafg nice catch!

**Pull request merged**